src/FragmentMakers.cpp

```diff
@@ -0,0 +1,80 @@
+#include "trigger/FragmentMakers.hpp"
+#include "dataformats/trigger/TriggerPrimitivesFragment.hpp"
+#include "triggeralgs/TriggerPrimitive.hpp"
+
+namespace dunedaq::trigger {
+
+//======================================================================
+std::unique_ptr<dataformats::Fragment>
+make_fragment(std::vector<TPSet>& tpsets)
+{
+  std::vector<triggeralgs::TriggerPrimitive> tps;
+  for (auto const& tpset : tpsets) {
+    for (auto const& tp : tpset.objects) {
+      tps.push_back(tp);
+    }
+  }
+
+  return make_fragment(tps);
+}
+
+//======================================================================
+std::unique_ptr<dataformats::Fragment>
+make_fragment(std::vector<triggeralgs::TriggerPrimitive>& tps)
+{
+  size_t num_bytes = sizeof(dataformats::TriggerPrimitivesFragment) +
+                   tps.size() * sizeof(dataformats::TriggerPrimitivesFragment::TriggerPrimitive);
+  std::unique_ptr<uint8_t[]> buffer(new uint8_t[num_bytes]);
+  dataformats::TriggerPrimitivesFragment* tpf = reinterpret_cast<dataformats::TriggerPrimitivesFragment*>(buffer.get());
+  tpf->version = dataformats::TriggerPrimitivesFragment::s_tpf_version;
+  tpf->num_trigger_primitives = tps.size();
+  size_t counter = 0;
+  for (auto const& tp : tps) {
+    dataformats::TriggerPrimitivesFragment::TriggerPrimitive& fragment_tp = tpf->at(counter++);
+    fragment_tp.time_start = tp.time_start;
+    fragment_tp.time_peak = tp.time_peak;
+    fragment_tp.time_over_threshold = tp.time_over_threshold;
+    fragment_tp.channel = tp.channel;
+    fragment_tp.adc_integral = tp.adc_integral;
+    fragment_tp.adc_peak = tp.adc_peak;
+    fragment_tp.detid = tp.detid;
+    fragment_tp.type = static_cast<uint32_t>(tp.type);
+    fragment_tp.algorithm = static_cast<uint32_t>(tp.algorithm);
+    fragment_tp.version = tp.version;
+    fragment_tp.flag = tp.flag;
+  }
+
+  std::unique_ptr<dataformats::Fragment> frag = std::make_unique<dataformats::Fragment>(buffer.get(), num_bytes);
```

Comment:
Would it be better to use the Fragment's piecewise constructor here?

```CXX
TriggerPrimitivesFragment f;
f.num_trigger_primitives = tps.size();
std::vector<dataformats::TriggerPrimitivesFragment::TriggerPrimitive> fragment_tps(tps.size());
for() {... same as before ... }
std::unique_ptr<dataformats::Fragment frag = std::make_unique<dataformats::Fragment>({{ &f, sizeof(f) }, { fragment_tps.data(), sizeof(dataformats::TriggerPrimitivesFragment::TriggerPrimitive) * tps.size() }});
```

Reply:
I guess "same as before" is incorrect, the initialization of fragment_tp would be deleted and there would be a `fragment_tps[counter++] = fragment_tp;`  at the end...